File: examples/elixir/user-server/test/user_service_router_test.exs
defmodule UserServiceRouterTest do
  use ExUnit.Case, async: true

  alias RuntimeTypes.HttpRequest

  @moduledoc """
  Request flow: %HttpRequest{} -> UserServiceRouter.dispatch/2 ->
  UserServiceRestJson1.decode_get_user_request/2 -> Handler.handle_get_user/3.
  Pass UserServiceServer as the handler module to invoke the generated server stub.
  """

  defp get_user_request do
    %HttpRequest{
      method: "GET",
      path: "/users/u-1",
      query: %{},
      headers: [],
      body: ""
    }
  end

  test "dispatch routes list users before user id path" do
    request = %HttpRequest{
      method: "GET",
      path: "/users",
      query: %{},
      headers: [],
      body: ""
    }

    assert {:error, :not_implemented} == UserServiceRouter.dispatch(UserServiceServer, request)
  end

  test "dispatch not found for extra user path segments" do
    request = %HttpRequest{
      method: "GET",
      path: "/users/u-1/extra",
      query: %{},
      headers: [],
      body: ""
    }

    assert {:error, {:not_found, "GET", "/users/u-1/extra"}} ==
             UserServiceRouter.dispatch(UserServiceServer, request)
  end

  test "dispatch routes to UserServer stub handler" do
    assert {:error, :not_implemented} == UserServiceRouter.dispatch(UserServiceServer, get_user_request())
  end

  test "dispatch decodes wire request before handler" do
    assert {:ok, out} = UserServiceRouter.dispatch(UserServiceServerProbe, get_user_request())
    assert out.user == %{"userId" => "u-1"}
  end

  test "dispatch not found for unknown route" do
    request = %{get_user_request() | method: "POST"}

    assert {:error, {:not_found, "POST", "/users/u-1"}} ==
             UserServiceRouter.dispatch(UserServiceServer, request)
  end

  test "dispatch routes create user" do
    request = %HttpRequest{
      method: "POST",
      path: "/users",
      query: %{},
      headers: [{"Content-Type", "application/json"}],
      body: Jason.encode!(%{"email" => "a@example.com"})
    }

    assert {:error, :not_implemented} == UserServiceRouter.dispatch(UserServiceServer, request)
  end
end

File: examples/erlang/user-server/test/user_service_router_test.erl
-module(user_service_router_test).

-include_lib("eunit/include/eunit.hrl").
-include("user_service_types.hrl").
-include("http_types.hrl").

%% Request flow: http_request -> user_service_router:dispatch/2 -> codec decode -> Handler:handle_<op>/3.
%% Pass user_service_server as Handler to invoke the generated server stub.

dispatch_routes_list_users_before_user_id_path_test() ->
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/users">>,
        query = #{},
        headers = [],
        body = <<>>
    },
    ?assertEqual({error, not_implemented}, user_service_router:dispatch(user_service_server, Req)).

dispatch_not_found_for_extra_user_path_segments_test() ->
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/users/u-1/extra">>,
        query = #{},
        headers = [],
        body = <<>>
    },
    ?assertEqual(
        {error, {not_found, <<"GET">>, <<"/users/u-1/extra">>}},
        user_service_router:dispatch(user_service_server, Req)
    ).

dispatch_routes_to_user_server_handler_test() ->
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/users/u-1">>,
        query = #{},
        headers = [],
        body = <<>>
    },
    ?assertEqual(
        {error, not_implemented},
        user_service_router:dispatch(user_service_server, Req)
    ).

dispatch_decodes_wire_request_before_handler_test() ->
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/users/u-1">>,
        query = #{},
        headers = [],
        body = <<>>
    },
    {ok, Out} = user_service_router:dispatch(user_service_server_probe, Req),
    ?assertEqual(#{<<"userId">> => <<"u-1">>}, Out#get_user_output.user).

dispatch_not_found_for_unknown_route_test() ->
    Req = #http_request{
        method = <<"POST">>,
        path = <<"/users/u-1">>,
        query = #{},
        headers = [],
        body = <<>>
    },
    ?assertEqual(
        {error, {not_found, <<"POST">>, <<"/users/u-1">>}},
        user_service_router:dispatch(user_service_server, Req)
    ).

dispatch_routes_create_user_test() ->
    Req = #http_request{
        method = <<"POST">>,
        path = <<"/users">>,
        query = #{},
        headers = [{<<"Content-Type">>, <<"application/json">>}],
        body = jsone:encode(#{<<"email">> => <<"a@example.com">>})
    },
    ?assertEqual({error, not_implemented}, user_service_router:dispatch(user_service_server, Req)).
